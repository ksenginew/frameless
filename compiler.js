let SERVERRE = /<\s*script(\s+[^]*?)?\sserver(\s+[^]*?)?>([^]*?)<\/\s*script\s*>/g

export function compiler(source) {
  let server = ''
  let static = source.replace(SERVERRE, (_, $1, $2, $3) => {
    server += $2 + '\n'
    return ''
  })
  return { server, static }
}

export function ssr(server, static) {
  return new Function(${server}return${static};)
}
