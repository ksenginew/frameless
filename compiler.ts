let RE =
  /<script([^]*?)>([^]*?)<\/script\s*>|<style([^]*?)>([^]*?)<\/style\s*>/g;

export function compiler(source: string, id = Math.random().toString(36).slice(2)) {
  let server = "";
  let style = "";
  let html = source.replace(RE, (_, $1, $2, $3, $4) => {
    if ($2)
      server += $2 + "\n";
    if ($4)
      style += $4;
    return "";
  });
  style.replace(/(?:\\,|[^,])+/g, selector => selector.replace(/(?<!\\)(?=\.)|$/, '.' + id));
  return { server, html, style };
}

export function ssrJs(server: string, _static: string) {
  return new Function("__", server + "return `" + _static + "`;");
}
