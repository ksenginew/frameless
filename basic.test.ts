import glob from 'glob';
import {compiler, ssrJs} from './compiler';
import {readFileSync} from 'fs'

describe('test', () => {
  test('all', () => {
    glob.sync('examples/**/App.html').map(file => {
      let { server, html, style } = compiler(readFileSync(file).toString())
      expect(`
      // file: ${file}
      function App() {
        ${server}

        return \`${html}\`
      }
      `).toMatchSnapshot()
    })
  })
})
