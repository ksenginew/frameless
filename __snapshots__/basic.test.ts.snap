// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test all 1`] = `
"
// file: examples/01-introduction/01-basics/App.html
------------------------------
js   = 
------------------------------
css  = 
------------------------------
html = <h1 class=’3adzgapv9bc’>Hello World!</h1>

"
`;

exports[`test all 2`] = `
"
// file: examples/01-introduction/02-adding-data/App.html
------------------------------
js   = 
	let name = 'world';


------------------------------
css  = 
------------------------------
html = 

<h1 class=’jss62dx38da’>Hello \${name.toUpperCase()}!</h1>

"
`;

exports[`test all 3`] = `
"
// file: examples/01-introduction/03-dynamic-attributes/App.html
------------------------------
js   = 
	let src = '/tutorial/image.gif';
	let name = 'Rick Astley';


------------------------------
css  = 
------------------------------
html = 

<img class=’hboe0t8rpo’ \${{src}} alt=’\${name} dances.’>

"
`;

exports[`test all 4`] = `
"
// file: examples/01-introduction/04-styling/App.html
------------------------------
js   = 
------------------------------
css  = 
	p.9pambbawi45 }
		color: purple;
		font-family: 'Comic Sans MS', cursive;
		font-size: 2em;
	}

------------------------------
html = <p class=’9pambbawi45’>This is a paragraph.</p>



"
`;
